### PR TITLE
adding refresh-token-endpoint

### PR DESCRIPTION
**Pull request by eapg**
* Adding refresh token endpoint and controller
* Adding tests for refresh token scenarios:
   * refresh unexpired access token should return same access token
   * refresh expired access token should return a new access token when refresh token has not expired
   * refresh expired access token when refresh token expired should return an Token expired error.

### CODE REVIEW

File: lib/ea_restaurant_data_loader_web/controllers/oauth2_controller.ex [hunk 1]
```diff
@@ -21,4 +21,21 @@ defmodule EaRestaurantDataLoaderWeb.Controllers.Oauth2Controller do
 
     send_resp(conn, :ok, login_json_response)
   end
+
+  def refresh_token(conn, _params) do
+    access_token = conn.body_params["access_token"]
+    refresh_token = conn.body_params["refresh_token"]
+    client_id = conn.body_params["client_id"]
+    client_secret = conn.body_params["client_secret"]
```
Comment (reynoldmorel) [inline lib/ea_restaurant_data_loader_web/controllers/oauth2_controller.ex]: You can refactor this:
```
%{
"access_token" => access_token,
"refresh_token" => refresh_token. 
"client_id" => client_id, 
"client_secret" => client_secret
} = conn.body_params
```

File: lib/ea_restaurant_data_loader_web/controllers/oauth2_controller.ex [hunk 2]
```diff
@@ -21,4 +21,21 @@ defmodule EaRestaurantDataLoaderWeb.Controllers.Oauth2Controller do
...
+
+    {_, refresh_token_json_response} =
+      try do
+        Oauth2Service.refresh_token(refresh_token, access_token, client_id, client_secret)
+        |> ApplicationUtil.parse_to_json()
+      rescue
+        _ in MatchError -> raise TokenExpiredError
```
Comment (reynoldmorel) [inline lib/ea_restaurant_data_loader_web/controllers/oauth2_controller.ex]: Not all errors are meant to be `TokenExpiredError` I don't think a `Matcherror` should be a `TokenExpiredError`

File: test/ea_restaurant_data_loader_web/controllers/oauth2_controller_test.exs [hunk 3]
```diff
@@ -49,5 +53,178 @@ defmodule EaRestaurantDataLoader.Test.EaRestaurantDataLoaderWeb.Controllers.Oaut
 
       assert_raise(BadRequest, ~r/bad request/, fn -> post(conn, "/login") end)
     end
+
+    test "refresh token - refresh unexpired access token", %{conn: conn} do
```
Comment (reynoldmorel) [inline test/ea_restaurant_data_loader_web/controllers/oauth2_controller_test.exs]: The description of the test should start saying `test "should ...."`, i.e. `test "should not refresh a valid token"`

**Review by reynoldmorel — CHANGES_REQUESTED** [review]

File: lib/ea_restaurant_data_loader_web/controllers/oauth2_controller.ex (hunk 1 above)
Comment (eapg) [reply lib/ea_restaurant_data_loader_web/controllers/oauth2_controller.ex]: Done.

File: lib/ea_restaurant_data_loader_web/controllers/oauth2_controller.ex (hunk 2 above)
Comment (eapg) [reply lib/ea_restaurant_data_loader_web/controllers/oauth2_controller.ex]: Done. i refactored the refresh token service to raise expired token intead of matching with {:ok,_}

File: test/ea_restaurant_data_loader_web/controllers/oauth2_controller_test.exs (hunk 3 above)
Comment (eapg) [reply test/ea_restaurant_data_loader_web/controllers/oauth2_controller_test.exs]: Done.

File: test/ea_restaurant_data_loader_web/controllers/oauth2_controller_test.exs [hunk 4]
```diff
@@ -49,5 +53,178 @@ defmodule EaRestaurantDataLoader.Test.EaRestaurantDataLoaderWeb.Controllers.Oaut
...
+      }
+
+      conn = post(conn, "/refresh_token", body_params)
+      {_, resp_body_decoded} = ApplicationUtil.decode_json(conn.resp_body)
+
+      assert conn.status == 200
+      assert resp_body_decoded["access_token"] != expired_access_token
+    end
+
+    test "should raise TokenExpiredError when refresh access token with expired refresh token", %{conn: conn} do
```
Comment (reynoldmorel) [inline test/ea_restaurant_data_loader_web/controllers/oauth2_controller_test.exs]: Typo: `Should ...`

**Review by reynoldmorel — APPROVED** [review]

File: test/ea_restaurant_data_loader_web/controllers/oauth2_controller_test.exs (hunk 4 above)
Comment (eapg) [reply test/ea_restaurant_data_loader_web/controllers/oauth2_controller_test.exs]: Done.

**Pull request merged**